### DISCUSSION TITLE
PSR-15 compatibility (request handlers / middleware)

### DISCUSSION THREAD

**Discussion by SimonFrings** [clue/framework-x#185] [Ideas]
PSR-15 compatibility would be a useful addition to Framework X. Not only will this introduce lots of interoperability with most middleware (or other projects) out there, it will also help us in our doing when it comes to new features for Framework X. Not every addition to the project needs to be integrated in X, in some cases it makes more sense to offer separate solutions. This way we don't overload the project itself.

Curious what you think about this addition!

**Comment by drupol** [comment]
I'm all for it.

**Comment by pfk84** [comment]
Is there a downside? Just asking because I suppose there was a reason why it wasn't done in the first place and the current implementation looks very similar.

**Reply by clue** [reply]
The main reason why we didn't implement PSR-15 from the get go is that PSR-15 is designed as a synchronous interface that does not take async responses into account. I understand why this was done and don't think this is necessarily a bad thing, but this definitely used to limit how you could employ this in async applications. Thanks to full fiber support via #168 and #128 released just a few weeks ago, it's now finally feasible to integrate full PSR-15 support also into async applications! :tada:

**Comment by clue** [comment]
Agree that PSR-15 compatibility would be a fantastic addition for Framework X! :100: It's a perfect fit for its design and I'm especially looking forward to being able to reuse all the existing middleware implementations from the entire PHP ecosystem!

Semantically, PSR-15 defines the exact same interfaces defined by Framework X, except that Framework X also support async responses using promises and coroutines. Now that full fiber support has been merged via #168 and #128, full PSR-15 compatibility is finally feasible.

Support for final request handlers (controllers) should be pretty straight forward, as it's only really a matter of calling its `handle()` method instead of requiring a `callable` instance.

Support for middleware request handlers is going to require a bit of work as the `$next` handler might return an async response in Framework X and as such does not currently implement the `RequestHandlerInterface`. Thanks to fibers, we can now map these interfaces, but this is definitely going to involve a bit of work.

TLDR: Definitely enjoys priority and would love to have this! There's hope this discussion can help us prioritize outstanding feature requests and steer project development.

**Reply by drupol** [reply]
Oooh yes :)

**Comment by dhrrgn** [comment]
I haven't actually tested this yet, but [for/http-middleware-psr15-adapter](https://github.com/friends-of-reactphp/http-middleware-psr15-adapter) should work fine with Framework X. Annoying to wrap each Middleware, but a first party solution using the same techniques should make a simple solution.

**Reply by clue** [reply]
Good pointer, https://github.com/friends-of-reactphp/http-middleware-psr15-adapter should indeed do exactly what is supposed to be built into Framework X in the future, so it might be a good stopgap solution! Future integration would be seamless, so you would be able to use callable middleware and PSR-15 middleware interchangeably.

**Comment by bwaidelich** [comment]
I would like to promote framework-x with a open API library of mine, but my integration is a PSR-15 middleware.
Would you be fine with me opening a PR that integrates the adapter to the core?

My (maybe naive) approach would be to simply tweak the `MiddlewareHandler` constructor along the lines of:

```php
class MiddlewareHandler
{
    /** @var list<callable> $handlers */
    private $handlers;

    /** @param list<callable|MiddlewareInterface> $handlers */
    public function __construct(array $handlers)
    {
        assert(count($handlers) >= 2);

        $this->handlers = array_map(function ($handler) {
            return $handler instanceof MiddlewareInterface ? new PsrMiddlewareAdapter($handler) : $handler;
         }, $handlers);
    }
```

But I was wondering whether that adapter needs the `async`/`await` logic as suggested by the adapter from friends-of-reactphp or whether it could be as simple as the one [suggested by me two weeks ago](https://github.com/clue/framework-x/discussions/130#discussioncomment-8630319) since it's probably blocking anyways!?

**Reply by SimonFrings** [reply]
> I would like to promote framework-x with a open API library of mine, but my integration is a PSR-15 middleware.
Would you be fine with me opening a PR that integrates the adapter to the core?

@bwaidelich Really awesome that you want to promote Framework X! We're always happy when people contribute to our projects and create good quality pull requests :heart:

The most important part for us is that the pull request should not only focus on the implementation side of things, which is most often the "easy" part. We also have to make sure that any new additions are also covered by our test suite (unit and/or acceptance tests) to confirm the suggested changes work as expected. The new feature also needs to be explained to our users, so we would require additional documentation for the `MiddlewareHandler` that describes when to use it and a recommendation for common use case(s).

If you're up for it, feel free to submit a pull request and I'd be happy to give it a review :+1:

**Reply by bwaidelich** [reply]
@SimonFrings there you go: https://github.com/clue/framework-x/pull/253
